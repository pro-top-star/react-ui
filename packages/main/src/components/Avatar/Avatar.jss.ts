import { CSSProperties } from 'react';
import { JSSTheme } from '../../interfaces/JSSTheme';

const size = (s): CSSProperties => ({
  height: s,
  width: s,
  minWidth: s,
  maxWidth: s,
  lineHeight: s,
  textAlign: 'center'
});

const styles = ({ parameters }: JSSTheme) => ({
  // outer flexbox
  avatar: {
    backgroundColor: parameters.sapAccentColor7,
    color: parameters.sapUiContentContrastTextColor,
    fontFamily: parameters.sapUiFontFamily,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundSize: 'cover',
    backgroundPosition: 'center',
    borderRadius: '0.25rem',
    '--sapUiContentNonInteractiveIconColor': parameters.sapContent_ContrastIconColor
  },
  // borderRadius 100%
  circle: {
    borderRadius: '100%'
  },
  // biggest avatar; 7rem
  sizeXL: {
    ...size('7rem'),
    fontSize: '3rem',
    '& ui5-icon': {
      width: '3rem',
      height: '3rem'
    }
  },
  // 5rem
  sizeL: {
    ...size('5rem'),
    fontSize: '2.25rem',
    '& ui5-icon': {
      width: '2.25rem',
      height: '2.25rem'
    }
  },
  // 4rem
  sizeM: {
    ...size('4rem'),
    fontSize: '1.5rem',
    '& ui5-icon': {
      width: '1.5rem',
      height: '1.5rem'
    }
  },
  // 3rem
  sizeS: {
    ...size('3rem'),
    fontSize: '1.125rem',
    '& ui5-icon': {
      width: '1.125rem',
      height: '1.125rem'
    }
  },
  // 2rem
  sizeXS: {
    ...size('2rem'),
    fontSize: '1rem',
    '& ui5-icon': {
      width: '1rem',
      height: '1rem'
    }
  },
  sizeCustom: ({ customFontSize, customDisplaySize }) => ({
    ...size(customDisplaySize),
    fontSize: customFontSize,
    '& ui5-icon': {
      width: customFontSize,
      height: customFontSize
    }
  })
});

export default styles;
