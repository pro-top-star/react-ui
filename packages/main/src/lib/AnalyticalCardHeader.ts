import { AnalyticalCardHeader } from '../components/AnalyticalCard/header/AnalyticalCardHeader';

export { AnalyticalCardHeader };
