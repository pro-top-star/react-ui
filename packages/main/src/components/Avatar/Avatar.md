```jsx
import { Avatar } from '@ui5/webcomponents-react/lib/Avatar';
```
