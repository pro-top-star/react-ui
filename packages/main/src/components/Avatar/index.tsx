import { Event } from '@ui5/webcomponents-react-base/lib/Event';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import { AvatarShape } from '@ui5/webcomponents-react/lib/AvatarShape';
import { AvatarSize } from '@ui5/webcomponents-react/lib/AvatarSize';
import React, { CSSProperties, FC, forwardRef, Ref, useCallback } from 'react';
import { createUseStyles } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';
import { JSSTheme } from '../../interfaces/JSSTheme';
import styles from './Avatar.jss';

export interface AvatarPropTypes extends CommonProps {
  size?: AvatarSize;
  shape?: AvatarShape;
  initials?: string;
  image?: string;
  onClick?: (event: Event) => void;
  children?: JSX.Element;
  customDisplaySize?: CSSProperties['width'];
  customFontSize?: CSSProperties['width'];
}

const useStyles = createUseStyles<JSSTheme, keyof ReturnType<typeof styles>>(styles, { name: 'Avatar' });

/**
 * <code>import { Avatar } from '@ui5/webcomponents-react/lib/Avatar';</code>
 */
const Avatar: FC<AvatarPropTypes> = forwardRef((props: AvatarPropTypes, ref: Ref<HTMLSpanElement>) => {
  const {
    initials,
    size,
    shape,
    image,
    onClick,
    customFontSize,
    customDisplaySize,
    children,
    className,
    style,
    tooltip,
    slot
  } = props;

  const classes = useStyles();

  const cssClasses = [classes.avatar];
  const inlineStyle: CSSProperties = {};
  if (size === AvatarSize.Custom) {
    inlineStyle.fontSize = customFontSize;
    inlineStyle.width = customDisplaySize;
    inlineStyle.height = customDisplaySize;
    inlineStyle.lineHeight = customDisplaySize;
  } else {
    cssClasses.push(classes[`size${size}`]);
  }

  if (shape === AvatarShape.Circle) {
    cssClasses.push(classes.circle);
  }

  if (image) {
    inlineStyle.backgroundImage = `url(${image})`;
  }

  if (onClick) {
    inlineStyle.cursor = 'pointer';
  }

  if (className) {
    cssClasses.push(className);
  }
  if (style) {
    Object.assign(inlineStyle, style);
  }

  const handleKeyDown = useCallback(
    (e) => {
      if (e.key === 'Enter') {
        onClick?.(Event.of(null, e));
      }
    },
    [onClick]
  );

  const handleOnClick = useCallback(
    (e) => {
      onClick?.(Event.of(null, e));
    },
    [onClick]
  );

  const passThroughProps = usePassThroughHtmlProps(props);

  return (
    <span
      ref={ref}
      className={cssClasses.join(' ')}
      style={inlineStyle}
      onClick={handleOnClick}
      tabIndex={0}
      onKeyDown={handleKeyDown}
      title={tooltip}
      slot={slot}
      {...passThroughProps}
    >
      {initials ? initials : children}
    </span>
  );
});

Avatar.defaultProps = {
  size: AvatarSize.S,
  shape: AvatarShape.Circle,
  initials: null,
  image: null,
  onClick: null,
  customDisplaySize: '3rem',
  customFontSize: '1.125rem'
};

Avatar.displayName = 'Avatar';

export { Avatar };
