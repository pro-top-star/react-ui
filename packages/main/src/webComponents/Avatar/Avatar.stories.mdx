import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs/blocks';
import { Avatar } from '@ui5/webcomponents-react/lib/Avatar';
import { AvatarBackgroundColor } from '@ui5/webcomponents-react/lib/AvatarBackgroundColor';
import '@ui5/webcomponents-icons/dist/employee.js';
import { AvatarFitType } from '@ui5/webcomponents-react/lib/AvatarFitType';
import { AvatarShape } from '@ui5/webcomponents-react/lib/AvatarShape';
import { AvatarSize } from '@ui5/webcomponents-react/lib/AvatarSize';
import { createSelectArgTypes } from '@shared/stories/createSelectArgTypes';
import { DocsHeader } from '@shared/stories/DocsHeader';
import { DocsCommonProps } from '@shared/stories/DocsCommonProps';

<Meta
  title="Data Display / Avatar"
  component={Avatar}
  argTypes={{
    ...createSelectArgTypes({
      backgroundColor: AvatarBackgroundColor,
      imageFitType: AvatarFitType,
      shape: AvatarShape,
      size: AvatarSize
    }),
    ...DocsCommonProps
  }}
  args={{
    backgroundColor: AvatarBackgroundColor.Accent6,
    icon: 'employee',
    imageFitType: AvatarFitType.Cover,
    shape: AvatarShape.Circle,
    size: AvatarSize.S,
    style: {},
    className: '',
    tooltip: '',
    slot: '',
    ref: null
  }}
/>

<DocsHeader />

<br />

## Example

<Canvas>
  <Story name="Default">
    {(args) => {
      return <Avatar {...args} />;
    }}
  </Story>
</Canvas>

## Properties

<ArgsTable story="." />
