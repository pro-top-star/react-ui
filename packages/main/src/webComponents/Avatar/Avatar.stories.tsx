import { select, text } from '@storybook/addon-knobs';
import { Avatar } from '@ui5/webcomponents-react/lib/Avatar';
import { AvatarBackgroundColor } from '@ui5/webcomponents-react/lib/AvatarBackgroundColor';
import { AvatarFitType } from '@ui5/webcomponents-react/lib/AvatarFitType';
import { AvatarShape } from '@ui5/webcomponents-react/lib/AvatarShape';
import { AvatarSize } from '@ui5/webcomponents-react/lib/AvatarSize';
import React from 'react';

export default {
  title: 'UI5 Web Components / Avatar',
  component: Avatar
};

export const generatedDefaultStory = () => (
  <Avatar
    backgroundColor={select('backgroundColor', AvatarBackgroundColor, AvatarBackgroundColor.Accent6)}
    icon={text('icon', 'employee')}
    image={text('image', '')}
    imageFitType={select('imageFitType', AvatarFitType, AvatarFitType.Cover)}
    initials={text('initials', '')}
    shape={select('shape', AvatarShape, AvatarShape.Circle)}
    size={select('size', AvatarSize, AvatarSize.S)}
  />
);

generatedDefaultStory.story = {
  name: 'Generated default story'
};
