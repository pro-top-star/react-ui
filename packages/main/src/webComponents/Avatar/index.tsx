import { AvatarBackgroundColor } from '@ui5/webcomponents-react/lib/AvatarBackgroundColor';
import { AvatarFitType } from '@ui5/webcomponents-react/lib/AvatarFitType';
import { AvatarShape } from '@ui5/webcomponents-react/lib/AvatarShape';
import { AvatarSize } from '@ui5/webcomponents-react/lib/AvatarSize';
import { withWebComponent, WithWebComponentPropTypes } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents/dist/Avatar';
import { FC } from 'react';

export interface AvatarPropTypes extends WithWebComponentPropTypes {
  /**
   * Defines the text alternative of the <code>ui5-avatar</code>. If not provided a default text alternative will be set, if present.
   */
  accessibleName?: string;
  /**
   * Defines the background color of the desired image. <br><br> Available options are: <ul> <li><code>Accent1</code></li> <li><code>Accent2</code></li> <li><code>Accent3</code></li> <li><code>Accent4</code></li> <li><code>Accent5</code></li> <li><code>Accent6</code></li> <li><code>Accent7</code></li> <li><code>Accent8</code></li> <li><code>Accent9</code></li> <li><code>Accent10</code></li> <li><code>Placeholder</code></li> <ul>
   */
  backgroundColor?: AvatarBackgroundColor;
  /**
   * Defines the name of the UI5 Icon, that would be displayed. <br> <b>Note:</b> If <code>image</code> is set, the property would be ignored. <br> <b>Note:</b> You should import the desired icon first, then use its name as "icon". <br><br> import "@ui5/webcomponents-icons/dist/icons/{icon_name}.js" <br> <pre>&lt;ui5-avatar icon-src="employee"></pre><br/><br/>
   *
   * See all the available icons in the <ui5-link target="_blank" href="https://openui5.hana.ondemand.com/test-resources/sap/m/demokit/iconExplorer/webapp/index.html" class="api-table-content-cell-link">Icon Explorer</ui5-link>.
   */
  icon?: string;
  /**
   * Defines the source path to the desired image.
   */
  image?: string;
  /**
   * Defines the fit type of the desired image. <br><br> Available options are: <ul> <li><code>Cover</code></li> <li><code>Contain</code></li> <ul>
   */
  imageFitType?: AvatarFitType;
  /**
   * Defines the displayed initials. <br> Up to two Latin letters can be displayed as initials in a <code>ui5-avatar</code>.
   */
  initials?: string;
  /**
   * Defines the shape of the <code>ui5-avatar</code>. <br><br> Available options are: <ul> <li><code>Circle</code></li> <li><code>Square</code></li> <ul>
   */
  shape?: AvatarShape;
  /**
   * Defines predefined size of the <code>ui5-avatar</code>. <br><br> Available options are: <ul> <li><code>XS</code></li> <li><code>S</code></li> <li><code>M</code></li> <li><code>L</code></li> <li><code>XL</code></li> <ul>
   */
  size?: AvatarSize;
}

/**
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/Avatar" target="_blank">UI5 Web Components Playground</a>
 */
const Avatar: FC<AvatarPropTypes> = withWebComponent<AvatarPropTypes>(
  'ui5-avatar',
  ['accessibleName', 'backgroundColor', 'icon', 'image', 'imageFitType', 'initials', 'shape', 'size'],
  [],
  [],
  []
);

Avatar.displayName = 'Avatar';

Avatar.defaultProps = {
  backgroundColor: AvatarBackgroundColor.Accent6,
  imageFitType: AvatarFitType.Cover,
  shape: AvatarShape.Circle,
  size: AvatarSize.S
};

export { Avatar };
