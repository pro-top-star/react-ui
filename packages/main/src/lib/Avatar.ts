import { Avatar } from '../webComponents/Avatar';

export { Avatar };
