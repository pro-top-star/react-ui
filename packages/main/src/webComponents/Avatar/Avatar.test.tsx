import { Avatar } from '@ui5/webcomponents-react/lib/Avatar';
import { mount } from 'enzyme';
import React from 'react';

describe('Avatar', () => {
  test('Basic Test (generated)', () => {
    const wrapper = mount(<Avatar />);
    expect(wrapper.render()).toMatchSnapshot();
  });
});
