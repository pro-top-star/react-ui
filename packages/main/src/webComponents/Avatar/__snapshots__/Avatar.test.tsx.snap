// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Basic Test (generated) 1`] = `
<DocumentFragment>
  <ui5-avatar
    image-fit-type="Cover"
    shape="Circle"
    ui5-avatar=""
  />
</DocumentFragment>
`;
